src/graph/sampler.cc

```diff
@@ -0,0 +1,487 @@
+      size_t pos = neigh_pos[idx_with_neigh].second;
+      CHECK_LT(pos, neighbor_list.size());
+      edge_size = neighbor_list[pos];
+      CHECK_LE(pos + edge_size * 2 + 1, neighbor_list.size());
+
+      std::copy_n(neighbor_list.begin() + pos + 1,
+                  edge_size,
+                  col_list_out + collected_nedges);
+      std::copy_n(neighbor_list.begin() + pos + edge_size + 1,
+                  edge_size,
+                  val_list_out + collected_nedges);
+      collected_nedges += edge_size;
+      idx_with_neigh++;
+    }
+    indptr_out[i+1] = indptr_out[i] + edge_size;
+  }
+
+  for (size_t i = 0; i < subg_csr->edge_ids.size(); i++)
+    subg_csr->edge_ids[i] = i;
+
+  if (neigh_type == "in")
+    subg.graph = GraphPtr(new ImmutableGraph(subg_csr, nullptr, IsMultigraph()));
+  else
+    subg.graph = GraphPtr(new ImmutableGraph(nullptr, subg_csr, IsMultigraph()));
+
+  return subg;
+}
+
+void CompactSubgraph(ImmutableGraph::CSR *subg,
+                     const std::unordered_map<dgl_id_t, dgl_id_t> &id_map) {
+  for (size_t i = 0; i < subg->indices.size(); i++) {
+    auto it = id_map.find(subg->indices[i]);
+    CHECK(it != id_map.end());
+    subg->indices[i] = it->second;
+  }
+}
+
+void ImmutableGraph::CompactSubgraph(IdArray induced_vertices) {
+  // The key is the old id, the value is the id in the subgraph.
+  std::unordered_map<dgl_id_t, dgl_id_t> id_map;
+  const dgl_id_t *vdata = static_cast<dgl_id_t *>(induced_vertices->data);
+  size_t len = induced_vertices->shape[0];
+  for (size_t i = 0; i < len; i++)
+    id_map.insert(std::pair<dgl_id_t, dgl_id_t>(vdata[i], i));
+  if (in_csr_)
+    dgl::CompactSubgraph(in_csr_.get(), id_map);
+  if (out_csr_)
+    dgl::CompactSubgraph(out_csr_.get(), id_map);
+}
+
+SampledSubgraph ImmutableGraph::NeighborUniformSample(IdArray seeds,
+                                                      const std::string &neigh_type,
+                                                      int num_hops, int expand_factor) const {
+  auto ret = SampleSubgraph(seeds,                 // seed vector
+                            nullptr,               // sample_id_probability
+                            neigh_type,
+                            num_hops,
+                            expand_factor);
+  std::static_pointer_cast<ImmutableGraph>(ret.graph)->CompactSubgraph(ret.induced_vertices);
+  return ret;
+}
+
+IdArray RandomWalk(
+    const GraphInterface *gptr,
+    IdArray seeds,
+    int num_traces,
+    int num_hops) {
+  const int num_nodes = seeds->shape[0];
+  const dgl_id_t *seed_ids = static_cast<dgl_id_t *>(seeds->data);
+  IdArray traces = IdArray::Empty(
+      {num_nodes, num_traces, num_hops + 1},
+      DLDataType{kDLInt, 64, 1},
+      DLContext{kDLCPU, 0});
+  dgl_id_t *trace_data = static_cast<dgl_id_t *>(traces->data);
+  unsigned int random_seed = time(nullptr);
+
+  // TODO: openmp support
```

Comment:
adding openmp should be easy.

Reply:
I'm not sure whether each thread should receive its own `random_seed`.

Reply:
every thread should have their own seeds to avoid write conflicts.